jigglypuffRL/deeprl/dqn/dqn.py

```diff
@@ -0,0 +1,239 @@
+        if network_type == "MLP":
+            self.model = MlpValue(
+                self.env.observation_space.shape[0],
+                self.env.action_space.n,
+                val_type="Qs",
+            )
+
+        if self.double_dqn and (network_type == "MLP"):
+            self.model = MlpValue(
+                self.env.observation_space.shape[0],
+                self.env.action_space.n,
+                val_type="Qs",
+            )
+            self.target_model = MlpValue(
+                self.env.observation_space.shape[0],
+                self.env.action_space.n,
+                val_type="Qs",
+            )
+
+        self.replay_buffer = ReplayBuffer(self.replay_size)
+        self.optimizer = opt.Adam(self.model.parameters(), lr=self.lr)
+
+    def update_target_model(self):
+        self.target_model.load_state_dict(self.model.state_dict())
+
+    def select_action(self, state, frame_idx):
+        epsilon = self.calculate_epsilon_by_frame(frame_idx)
+
+        if np.random.rand() > epsilon:
+            state = Variable(torch.FloatTensor(state))
+            q_value = self.model(state)
+            action = np.argmax(q_value.detach().numpy())
+        else:
+            action = self.env.action_space.sample()
+
+        return action
+
+    def get_td_loss(self):
+        state, action, reward, next_state, done = self.replay_buffer.sample(
+            self.batch_size
+        )
+
+        state = Variable(torch.FloatTensor(np.float32(state)))
+        next_state = Variable(torch.FloatTensor(np.float32(next_state)), volatile=True)
+        action = Variable(torch.LongTensor(action.long()))
+        reward = Variable(torch.FloatTensor(reward))
+        done = Variable(torch.FloatTensor(done))
+
+        q_values = self.model(state)
+        if self.double_dqn:
+            next_q_values = self.target_model(next_state)
+        else:
+            next_q_values = self.model(next_state)
+
+        q_value = q_values.gather(1, action.unsqueeze(1)).squeeze(1)
+        next_q_value = next_q_values.max(1)[0]
+        expected_q_value = reward + self.gamma * next_q_value * (1 - done)
+
+        loss = F.smooth_l1_loss(q_value, expected_q_value)
+
+        self.loss_hist.append(loss)
+
+        return loss
+
+    def update_params(self):
+        loss = self.get_td_loss()
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+
+    def calculate_epsilon_by_frame(self, frame_idx):
+        return self.min_epsilon + (self.max_epsilon - self.min_epsilon) * np.exp(
+            -1.0 * frame_idx / self.epsilon_decay
+        )
+
+    def plot(self):
+        plt.figure(figsize=(20, 5))
+        plt.subplot(131)
+        plt.title(
+            "Average reward for last 10 frames: %s" % (np.mean(self.reward_hist[-10:]))
+        )
+        plt.plot(self.reward_hist)
+        plt.subplot(132)
+        plt.title("loss")
+        plt.plot(self.loss_hist)
+        plt.show()
```

Comment:
Not required since we're already using tensorboard

Reply:
Cool will remove this was not using tensorboard, this was my sort of helper function.